pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,327 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package ebpf
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	manager "github.com/DataDog/ebpf-manager"
+	"github.com/cilium/ebpf"
+)
+
+const defaultBatchSize = 100
+
+var (
+	batchAPISupportedOnce sync.Once
+	batchAPISupported     bool
+)
+
+// BatchAPISupported returns true if the kernel supports the batch API for maps
+func BatchAPISupported() bool {
+	batchAPISupportedOnce.Do(func() {
+		// Do feature detection directly instead of based on kernel versions for more accuracy
+		m, err := ebpf.NewMap(&ebpf.MapSpec{
+			Type:       ebpf.Hash,
+			KeySize:    4,
+			ValueSize:  4,
+			MaxEntries: 10,
+		})
+		if err != nil {
+			log.Warnf("Failed to create map for batch API test: %v, will mark batch API as unsupported", err)
+			batchAPISupported = false
+			return
+		}
+
+		keys := make([]uint32, 1)
+		values := make([]uint32, 1)
+
+		// Do a batch update, check the result.
+		// We do an update instead of a lookup because it's more reliable for detection
+		_, err = m.BatchUpdate(keys, values, nil)
+		batchAPISupported = err == nil
+	})
+	return batchAPISupported
+}
+
+// GenericMap is a wrapper around ebpf.Map that allows to use generic types.
+// Also includes support for batch iterations
+type GenericMap[K interface{}, V interface{}] struct {
+	m *ebpf.Map
+}
+
+// NewGenericMap creates a new GenericMap with the given spec. Key and Value sizes are automatically
+// inferred from the types of K and V.
+// Important: if the map is a per-cpu map, V must be a slice type
+func NewGenericMap[K interface{}, V interface{}](spec *ebpf.MapSpec) (*GenericMap[K, V], error) {
+	// Automatic inference of sizes. We assume that K/V are simple types that
+	// can be instantiated with no arguments
+	var kval K
+	var vval V
+
+	err := validateValueTypeForMapType[V](spec.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	spec.KeySize = uint32(unsafe.Sizeof(kval))
+
+	if isPerCPU(spec.Type) {
+		spec.ValueSize = uint32(reflect.TypeOf(vval).Elem().Size())
+	} else {
+		spec.ValueSize = uint32(unsafe.Sizeof(vval))
+	}
+
+	m, err := ebpf.NewMap(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GenericMap[K, V]{
+		m: m,
+	}, nil
+}
+
+func validateValueTypeForMapType[V interface{}](t ebpf.MapType) error {
+	var vval V
+	if isPerCPU(t) && reflect.TypeOf(vval).Kind() != reflect.Slice {
+		return errors.New("per-cpu maps require a slice type for the value")
+	}
```

Comment:
it will be better to add the actual type we got here
```suggestion
	if isPerCPU(t) && reflect.TypeOf(vval).Kind() != reflect.Slice {
		return errors.New("per-cpu maps require a slice type for the value, instead got %T", vval)
	}
```